flink-streaming-java/src/test/java/org/apache/flink/streaming/runtime/tasks/RestoreStreamTaskTest.java

```diff
@@ -0,0 +1,325 @@
+import org.apache.flink.runtime.checkpoint.CheckpointMetaData;
+import org.apache.flink.runtime.checkpoint.CheckpointOptions;
+import org.apache.flink.runtime.checkpoint.OperatorSubtaskState;
+import org.apache.flink.runtime.checkpoint.StateAssignmentOperation;
+import org.apache.flink.runtime.checkpoint.TaskStateSnapshot;
+import org.apache.flink.runtime.jobgraph.OperatorID;
+import org.apache.flink.runtime.jobgraph.OperatorInstanceID;
+import org.apache.flink.runtime.operators.testutils.MockInputSplitProvider;
+import org.apache.flink.runtime.state.StateInitializationContext;
+import org.apache.flink.runtime.state.StateSnapshotContext;
+import org.apache.flink.streaming.api.operators.AbstractStreamOperator;
+import org.apache.flink.streaming.api.operators.OneInputStreamOperator;
+import org.apache.flink.streaming.runtime.streamrecord.StreamRecord;
+import org.apache.flink.streaming.util.TestHarnessUtil;
+import org.apache.flink.util.TestLogger;
+
+import org.junit.Test;
+
+import java.util.Collections;
+import java.util.Optional;
+import java.util.concurrent.ConcurrentLinkedQueue;
+import java.util.concurrent.atomic.AtomicLong;
+
+import static org.junit.Assert.assertEquals;
+
+/**
+ * Tests ensuring correct behaviour of {@link org.apache.flink.runtime.state.ManagedInitializationContext#isRestored}
+ * method.
+ */
+public class RestoreStreamTaskTest extends TestLogger {
+	@Test
+	public void testRestore() throws Exception {
+		AcknowledgeStreamMockEnvironment environment1 = processRecords(
+			new OperatorID(42L, 42L),
+			new CounterOperator(),
+			new OperatorID(44L, 44L),
+			new CounterOperator(),
+			Optional.empty());
+
+		assertEquals(2, environment1.getCheckpointStateHandles().getSubtaskStateMappings().size());
+
+		TaskStateSnapshot stateHandles = environment1.getCheckpointStateHandles();
+
+		try {
+			AcknowledgeStreamMockEnvironment environment2 = processRecords(
+				new OperatorID(42L, 42L),
+				new RestoreCounterOperator(),
+				new OperatorID(44L, 44L),
+				new RestoreCounterOperator(),
+				Optional.of(stateHandles));
+
+			assertEquals(2, RestoreCounterOperator.RESTORE_COUNTER.get());
+		}
+		finally {
+			RestoreCounterOperator.RESTORE_COUNTER.getAndSet(0);
+		}
+	}
+
+	@Test
+	public void testRestoreWithNewId() throws Exception {
+		AcknowledgeStreamMockEnvironment environment1 = processRecords(
+			new OperatorID(42L, 42L),
+			new CounterOperator(),
+			new OperatorID(44L, 44L),
+			new CounterOperator(),
+			Optional.empty());
+
+		assertEquals(2, environment1.getCheckpointStateHandles().getSubtaskStateMappings().size());
+
+		TaskStateSnapshot stateHandles = environment1.getCheckpointStateHandles();
+
+		try {
+			AcknowledgeStreamMockEnvironment environment2 = processRecords(
+				new OperatorID(4242L, 4242L),
+				new RestoreCounterOperator(),
+				new OperatorID(44L, 44L),
+				new RestoreCounterOperator(),
+				Optional.of(stateHandles));
+
+			assertEquals(1, RestoreCounterOperator.RESTORE_COUNTER.get());
```

Comment:
Couldn't this also be interpreted in the exact opposite way because we don't validate for which of the two operators we counted `isRestored == true`? This be correct in connection with the other tests in this class, but i think the invariant is not completely clear from only this test and it is not fully self-contained.

Reply:
Maybe keeping a set of the restored operator ids instead of a count already does the job?